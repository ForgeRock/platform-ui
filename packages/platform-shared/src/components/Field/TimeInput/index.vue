<!-- Copyright (c) 2021-2022 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <div class="position-relative">
    <FrInputLayout
      :id="id"
      :name="name"
      :description="description"
      :errors="errors"
      :is-html="isHtml"
      :label="label"
      :validation="validation"
      :validation-immediate="validationImmediate">
      <input
        v-model="inputValue"
        ref="input"
        type="time"
        :class="[{'is-invalid': errorMessages && errorMessages.length, 'polyfill-placeholder': label }, 'form-control']"
        :data-vv-as="label"
        :disabled="disabled"
        :id="id"
        :name="name"
        :readonly="readonly"
        @input="emitValidTime">
    </FrInputLayout>
    <BFormTimepicker
      v-model="inputValue"
      :dropleft="dropleft"
      button-only
      label-help=""
      no-close-button
      ref="timePicker"
      show-seconds
      :class="[{'is-invalid': errorMessages && errorMessages.length }, 'form-control time-button position-absolute']"
      :disabled="disabled"
      :id="id"
      :name="name"
      :aria-label="getTranslation(label)"
      @context="emitValidTime" />
  </div>
</template>

<script>
import { has, isEqual, cloneDeep } from 'lodash';
import { BFormTimepicker } from 'bootstrap-vue';
import TranslationMixin from '@forgerock/platform-shared/src/mixins/TranslationMixin';
import FrInputLayout from '../Wrapper/InputLayout';
import InputMixin from '../Wrapper/InputMixin';

/**
 * Input for time (HH:mm:ss.mmmZ) with a floating label in the center, this will move when a user types into the input.
 */
export default {
  name: 'TimeInput',
  mixins: [
    InputMixin,
    TranslationMixin,
  ],
  components: {
    BFormTimepicker,
    FrInputLayout,
  },
  props: {
    dropleft: {
      default: true,
      type: Boolean,
    },
  },
  methods: {
    /**
     * Emits out a formatted time in the form of HH:mm:ss or an empty string if time is invalid
     *
     * @param {String} selectedTime - The current time value selected
     * @emits {String} The fully formatted time string
     */
    emitValidTime(selectedTime) {
      const emitTime = has(selectedTime, 'value') ? selectedTime.value : selectedTime;
      if (emitTime) {
        this.$emit('input', `${`${emitTime}:00`.substring(0, 8)}.000`);
      } else {
        this.$emit('input', '');
      }
    },
    /**
     * Sets the time value for both the time picket and the time input
     */
    inputValueHandler(inputValue) {
      this.inputValue = inputValue.substring(0, 8);
    },
    /**
    * formats input value to be in the format that date picker and input expect
    *
    * @param {Number|String} newVal value to be set for input vale
    */
    setInputValue(newVal) {
      if (newVal !== undefined && newVal !== null && !isEqual(this.oldValue, newVal)) {
        this.inputValue = newVal.substring(0, 8);
        this.oldValue = cloneDeep(newVal);
      }
    },
  },
};
</script>

<style lang="scss" scoped>
  .time-button {
    width: 46.5px;
    border: none;
    right: 1px;
    top: 1px;
    height: calc(100% - 2px);

    /deep/ button {
      background-color: initial;
      color: $gray-500;
      padding: 0.85rem 0.5rem;
    }

    /deep/ &.show > .btn-secondary.dropdown-toggle {
      background-color: initial;
      color: $gray-500;
    }
  }

  .form-control.is-invalid {
    background-image: none;
  }
</style>
