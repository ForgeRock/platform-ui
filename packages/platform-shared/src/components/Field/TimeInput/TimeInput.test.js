/**
 * Copyright (c) 2021-2022 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { createLocalVue, shallowMount } from '@vue/test-utils';
import TimeInput from './index';

const localVue = createLocalVue();

const defaultMixinProps = {
  id: '',
  errorMessages: [],
  name: '',
  description: '',
  isHtml: false,
  label: '',
  readonly: false,
  autofocus: true,
};

describe('TimeInput', () => {
  it('TimeInput component loaded', () => {
    const wrapper = shallowMount(TimeInput, {
      localVue,
      propsData: {
        ...defaultMixinProps,
      },
      mocks: {
        $t: (text) => (text),
      },
    });

    expect(wrapper.name()).toBe('TimeInput');
  });

  it('builds a formatted time', () => {
    const wrapper = shallowMount(TimeInput, {
      localVue,
      propsData: {
        ...defaultMixinProps,
        label: 'test',
      },
      mocks: {
        $t: (text) => (text),
      },
    });

    wrapper.vm.emitValidTime('01:01');
    expect(wrapper.emitted().input[0][0]).toEqual('01:01:00.000');
    wrapper.vm.emitValidTime('02:02:02');
    expect(wrapper.emitted().input[1][0]).toEqual('02:02:02.000');
    wrapper.vm.emitValidTime('');
    expect(wrapper.emitted().input[2][0]).toEqual('');
    wrapper.setData({ inputValue: '03:03:03' });
    expect(wrapper.vm.inputValue).toEqual('03:03:03');
  });
});
