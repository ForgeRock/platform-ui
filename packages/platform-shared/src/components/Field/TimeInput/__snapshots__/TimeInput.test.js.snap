// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeInput builds a formatted time: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) Vue 3's render function API has changed. You can opt-in to the new API with:

  configureCompat({ RENDER_FUNCTION: false })

  (This can also be done per-component via the "compatConfig" option.)
  Details: https://v3-migration.vuejs.org/breaking-changes/render-function-api.html",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation WATCH_ARRAY) (2)",
    "
",
    " at <TimeInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation GLOBAL_EXTEND) (2)",
    "
",
    " at <TimeInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation COMPONENT_V_MODEL) v-model usage on component has changed in Vue 3. Component that expects to work with v-model should now use the "modelValue" prop and emit the "update:modelValue" event. You can update the usage and then opt-in to Vue 3 behavior on a per-component basis with \`compatConfig: { COMPONENT_V_MODEL: false }\`.
  Details: https://v3-migration.vuejs.org/breaking-changes/v-model.html",
    "
",
    " at <BFormTimepicker",
    "modelValue=""",
    "onUpdate:modelValue=fn",
    "dropleft=true",
    " ...",
    ">",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "isHtml" with v-bind value \`false\` will render isHtml="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <FrInputLayout",
    "id="floatingLabelInput3"",
    "name=""",
    "description=""",
    " ...",
    ">",
    "
",
    " at <TimeInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "readonlyLabel" with v-bind value \`false\` will render readonlyLabel="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <FrInputLayout",
    "id="floatingLabelInput3"",
    "name=""",
    "description=""",
    " ...",
    ">",
    "
",
    " at <TimeInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "showLengthCount" with v-bind value \`false\` will render showLengthCount="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <FrInputLayout",
    "id="floatingLabelInput3"",
    "name=""",
    "description=""",
    " ...",
    ">",
    "
",
    " at <TimeInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "disabled" with v-bind value \`false\` will render disabled="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormTimepicker",
    "dropleft=true",
    "button-only=""",
    "label-help=""",
    " ...",
    ">",
    "
",
    " at <TimeInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "dropright" with v-bind value \`false\` will render dropright="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormTimepicker",
    "dropleft=true",
    "button-only=""",
    "label-help=""",
    " ...",
    ">",
    "
",
    " at <TimeInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "dropup" with v-bind value \`false\` will render dropup="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormTimepicker",
    "dropleft=true",
    "button-only=""",
    "label-help=""",
    " ...",
    ">",
    "
",
    " at <TimeInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "hideHeader" with v-bind value \`false\` will render hideHeader="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormTimepicker",
    "dropleft=true",
    "button-only=""",
    "label-help=""",
    " ...",
    ">",
    "
",
    " at <TimeInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "noFlip" with v-bind value \`false\` will render noFlip="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormTimepicker",
    "dropleft=true",
    "button-only=""",
    "label-help=""",
    " ...",
    ">",
    "
",
    " at <TimeInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "nowButton" with v-bind value \`false\` will render nowButton="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormTimepicker",
    "dropleft=true",
    "button-only=""",
    "label-help=""",
    " ...",
    ">",
    "
",
    " at <TimeInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "required" with v-bind value \`false\` will render required="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormTimepicker",
    "dropleft=true",
    "button-only=""",
    "label-help=""",
    " ...",
    ">",
    "
",
    " at <TimeInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "resetButton" with v-bind value \`false\` will render resetButton="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormTimepicker",
    "dropleft=true",
    "button-only=""",
    "label-help=""",
    " ...",
    ">",
    "
",
    " at <TimeInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation ATTR_FALSE_VALUE) Attribute "right" with v-bind value \`false\` will render right="false" instead of removing it in Vue 3. To remove the attribute, use \`null\` or \`undefined\` instead. If the usage is intended, you can disable the compat behavior and suppress this warning with:

  configureCompat({ ATTR_FALSE_VALUE: false })

  Details: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html",
    "
",
    " at <BFormTimepicker",
    "dropleft=true",
    "button-only=""",
    "label-help=""",
    " ...",
    ">",
    "
",
    " at <TimeInput",
    "id=""",
    "errorMessages=",
    [],
    "name=""",
    " ...",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;
